Panic on slackdump.exe export due to "index out of range" error

When running slackdump.exe export with a data.txt file specifying channel IDs to include/exclude, the program crashes with a runtime error:

`slackdump.exe export @data.txt`
panic: runtime error: index out of range [0] with length 0

**To Reproduce**
Steps to reproduce the behavior:
1. Create a data.txt file that includes channel IDs to include and exclude in the export. For example:

C01A2B3C4D5
^C02X3Y4Z5W6

2- Run the command:  slackdump.exe export @data.txt

**Expected behavior**
The export process should run without errors, correctly processing the included and excluded channel IDs.

Actual Behavior
The program panics and outputs the following error:
**Output**
```
2024/12/13 19:59:42 INFO temporary directory in use tmpdir=C:\Users\test_user\AppData\Local\Temp\slackdump-528652615
.   [0s] 2024/12/13 19:59:42 INFO running export...
. <Thread[C02MAU7DZ8C:1730147000.804399]>  [10s] panic: runtime error: index out of range [0] with length 0

goroutine 57 [running]:
github.com/rusq/slackdump/v3/stream.procChanMsg({0x2172d98, 0xc000459680}, {0x2175db0, 0xc0001dda00}, 0xc000166ee0, 0xc0000cc680, 0x1, {0x2a4f840, 0x0, 0x0})
        /home/runner/work/slackdump/slackdump/stream/conversation.go:299 +0x76d
github.com/rusq/slackdump/v3/stream.(*Stream).channelWorker.func1({0x2a4f840?, 0xc0001685d0?, 0x1c26288?}, 0x1)
        /home/runner/work/slackdump/slackdump/stream/stream_workers.go:31 +0x7e
github.com/rusq/slackdump/v3/stream.(*Stream).channel(0xc0000908c0, {0x2172d98, 0xc0001685d0}, {0xc000497380, 0xb}, 0xc0009e5ea0)
        /home/runner/work/slackdump/slackdump/stream/conversation.go:202 +0x29e
github.com/rusq/slackdump/v3/stream.(*Stream).channelWorker(0xc0000908c0, {0x2172d98, 0xc000459680}, {0x2175db0, 0xc0001dda00}, 0xc000166f50, 0xc000166ee0, 0xc000166e70)
        /home/runner/work/slackdump/slackdump/stream/stream_workers.go:30 +0x345
github.com/rusq/slackdump/v3/stream.(*Stream).Conversations.func1()
        /home/runner/work/slackdump/slackdump/stream/conversation.go:76 +0x76
created by github.com/rusq/slackdump/v3/stream.(*Stream).Conversations in goroutine 76
        /home/runner/work/slackdump/slackdump/stream/conversation.go:74 +0x205
```
        

**Desktop (please complete the following information):**
 - OS: Windows 11
 - OS Version: 10.0.26100 Build 26100
 - Slackdump Version: Slackdump 3.0.0-alpha (commit: 14a4f89b6380387f91935f62c341db07c7c6c7b1) built on: 2024-11-24T03:48:45Z

**Additional context**
Add any other context about the problem here.

--- a/internal/chunk/dirproc/conversations.go
+++ b/internal/chunk/dirproc/conversations.go
@@ -124,7 +124,7 @@ func (cv *Conversations) ChannelInfo(ctx context.Context, ci *slack.Channel, thr
 	return r.ChannelInfo(ctx, ci, threadTS)
 }
 
-// Messages is called for each message that is retrieved.
+// Messages is called for each message slice that is retrieved.
 func (cv *Conversations) Messages(ctx context.Context, channelID string, numThreads int, isLast bool, mm []slack.Message) error {
 	ctx, task := trace.NewTask(ctx, "Messages")
 	defer task.End()

--- a/internal/chunk/dirproc/conversations_test.go
+++ b/internal/chunk/dirproc/conversations_test.go
@@ -143,6 +143,51 @@ func TestConversations_Messages(t *testing.T) {
 			},
 			wantErr: true,
 		},
+		{
+			name: "empty message slice, not last",
+			fields: fields{
+				subproc:     nil,
+				recordFiles: false,
+				tf:          nil,
+			},
+			args: args{
+				ctx:        textCtx,
+				channelID:  "channelID",
+				numThreads: 0,
+				isLast:     false,
+				mm:         []slack.Message{},
+			},
+			expectFn: func(mt *Mocktracker, mh *Mockdatahandler) {
+				mt.EXPECT().Recorder(gomock.Any()).Return(mh, nil)
+				mh.EXPECT().Messages(gomock.Any(), "channelID", 0, false, []slack.Message{}).Return(nil)
+				mh.EXPECT().Add(0).Return(0)
+			},
+			wantErr: false,
+		},
+		{
+			name: "empty message slice, last",
+			fields: fields{
+				subproc:     nil,
+				recordFiles: false,
+				tf:          nil,
+			},
+			args: args{
+				ctx:        textCtx,
+				channelID:  "channelID",
+				numThreads: 0,
+				isLast:     true,
+				mm:         []slack.Message{},
+			},
+			expectFn: func(mt *Mocktracker, mh *Mockdatahandler) {
+				mt.EXPECT().Recorder(gomock.Any()).Return(mh, nil)
+				mh.EXPECT().Messages(gomock.Any(), "channelID", 0, true, []slack.Message{}).Return(nil)
+				mh.EXPECT().Add(0).Return(1)
+				mh.EXPECT().Dec().Return(0)
+				mt.EXPECT().RefCount(chunk.ToFileID("channelID", "", false)).Return(0)
+				mt.EXPECT().Unregister(chunk.ToFileID("channelID", "", false)).Return(nil)
+			},
+			wantErr: false,
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

--- a/stream/conversation.go
+++ b/stream/conversation.go
@@ -2,7 +2,6 @@ package stream
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"log/slog"
 	"runtime/trace"
@@ -296,7 +295,10 @@ func procChanMsg(ctx context.Context, proc processor.Conversations, threadC chan
 		}
 	}
 	if err := proc.Messages(ctx, channel.ID, len(trs), isLast, mm); err != nil {
-		return 0, fmt.Errorf("failed to process message chunk starting with id=%s (size=%d): %w", mm[0].Msg.Timestamp, len(mm), err)
+		if len(mm) == 0 {
+			return 0, fmt.Errorf("channel %s: failed to process empty message chunk: %w", channel.ID, err)
+		}
+		return 0, fmt.Errorf("channel %s: failed to process message chunk starting with id=%s (size=%d): %w", channel.ID, mm[0].Msg.Timestamp, len(mm), err)
 	}
 	for _, tr := range trs {
 		threadC <- tr
@@ -305,17 +307,23 @@ func procChanMsg(ctx context.Context, proc processor.Conversations, threadC chan
 }
 
 func procThreadMsg(ctx context.Context, proc processor.Conversations, channel *slack.Channel, threadTS string, threadOnly bool, isLast bool, msgs []slack.Message) error {
-	// extract files from thread messages
+	lg := slog.With("channel_id", channel.ID, "thread_ts", threadTS, "is_last", isLast, "msg_count", len(msgs))
 	if len(msgs) == 0 {
-		return errors.New("empty messages slice")
-	}
-	if err := procFiles(ctx, proc, channel, msgs[1:]...); err != nil {
-		return err
+		lg.Debug("empty thread messages")
+		return nil
 	}
 	// slack returns the thread starter as the first message with every
-	// call, so we use it as a parent message.
-	if err := proc.ThreadMessages(ctx, channel.ID, msgs[0], threadOnly, isLast, msgs[1:]); err != nil {
-		return fmt.Errorf("failed to process thread message id=%s, thread_ts=%s: %w", msgs[0].Msg.Timestamp, threadTS, err)
+	// call, so we use it as a head message.
+	head, rest := msgs[0], []slack.Message{}
+	if len(msgs) > 1 {
+		rest = msgs[1:]
+	}
+	// extract files from thread messages
+	if err := procFiles(ctx, proc, channel, rest...); err != nil {
+		return err
+	}
+	if err := proc.ThreadMessages(ctx, channel.ID, head, threadOnly, isLast, rest); err != nil {
+		return fmt.Errorf("failed to process thread message id=%s, thread_ts=%s: %w", head.Msg.Timestamp, threadTS, err)
 	}
 	return nil
 }

--- a/stream/conversation_test.go
+++ b/stream/conversation_test.go
@@ -1,0 +1,133 @@
+package stream
+
+import (
+	"context"
+	"testing"
+
+	"github.com/rusq/slack"
+	"github.com/rusq/slackdump/v3/internal/fixtures"
+	"github.com/rusq/slackdump/v3/mocks/mock_processor"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/mock/gomock"
+)
+
+var TestChannel = &slack.Channel{
+	GroupConversation: slack.GroupConversation{
+		Conversation: slack.Conversation{
+			ID: "C12345678",
+		},
+	},
+}
+
+func Test_procChanMsg(t *testing.T) {
+	type args struct {
+		ctx context.Context
+		// proc    processor.Conversations // supplied by test
+		threadC chan<- request
+		channel *slack.Channel
+		isLast  bool
+		mm      []slack.Message
+	}
+	tests := []struct {
+		name     string
+		args     args
+		expectFn func(mp *mock_processor.MockConversations)
+		want     int
+		wantErr  bool
+	}{
+		{
+			"empty messages slice",
+			args{
+				context.Background(),
+				make(chan request),
+				TestChannel,
+				true,
+				[]slack.Message{},
+			},
+			func(mp *mock_processor.MockConversations) {
+				mp.EXPECT().Messages(gomock.Any(), TestChannel.ID, 0, true, []slack.Message{}).Times(1)
+			},
+			0,
+			false,
+		},
+		{
+			"empty message slice, processor error",
+			args{
+				context.Background(),
+				make(chan request),
+				TestChannel,
+				true,
+				[]slack.Message{},
+			},
+			func(mp *mock_processor.MockConversations) {
+				mp.EXPECT().Messages(gomock.Any(), TestChannel.ID, 0, true, []slack.Message{}).Return(assert.AnError).Times(1)
+			},
+			0,
+			true,
+		},
+		{
+			"non-empty messages slice",
+			args{
+				context.Background(),
+				make(chan request),
+				TestChannel,
+				true,
+				fixtures.Load[[]slack.Message](fixtures.TestChannelEveryoneMessagesNativeExport),
+			},
+			func(mp *mock_processor.MockConversations) {
+				mp.EXPECT().Messages(gomock.Any(), TestChannel.ID, 0, true, fixtures.Load[[]slack.Message](fixtures.TestChannelEveryoneMessagesNativeExport)).Times(1)
+				mp.EXPECT().Files(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+			},
+			0,
+			false,
+		},
+		{
+			"non-empty messages slice,files processor error",
+			args{
+				context.Background(),
+				make(chan request),
+				TestChannel,
+				true,
+				fixtures.Load[[]slack.Message](fixtures.TestChannelEveryoneMessagesNativeExport),
+			},
+			func(mp *mock_processor.MockConversations) {
+				mp.EXPECT().Files(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(assert.AnError)
+			},
+			0,
+			true,
+		},
+		{
+			"non-empty messages slice, messages processor error",
+			args{
+				context.Background(),
+				make(chan request),
+				TestChannel,
+				true,
+				fixtures.Load[[]slack.Message](fixtures.TestChannelEveryoneMessagesNativeExport),
+			},
+			func(mp *mock_processor.MockConversations) {
+				mp.EXPECT().Messages(gomock.Any(), TestChannel.ID, 0, true, fixtures.Load[[]slack.Message](fixtures.TestChannelEveryoneMessagesNativeExport)).Return(assert.AnError).Times(1)
+				mp.EXPECT().Files(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+			},
+			0,
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			mp := mock_processor.NewMockConversations(ctrl)
+			if tt.expectFn != nil {
+				tt.expectFn(mp)
+			}
+			got, err := procChanMsg(tt.args.ctx, mp, tt.args.threadC, tt.args.channel, tt.args.isLast, tt.args.mm)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("procChanMsg() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("procChanMsg() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}

